cli.js

```diff
@@ -24,82 +21,57 @@ const wait = (ms) => new Promise((resolve) => setTimeout(resolve, ms));
 
+  // User Interaction - https://tools.ietf.org/html/rfc8628#section-3.3
   await prompts({
     type: 'invisible',
     message: `Press any key to open up the browser to login or press ctrl-c to abort. You should see the following code: ${user_code}. It expires in ${expires_in % 60 === 0 ? `${expires_in / 60} minutes` : `${expires_in} seconds`}.`,
   });
-
   // opens the verification_uri_complete URL using the system-register handler for web links (browser)
   open(verification_uri_complete);
 
-  let done;
+  // Device Access Token Request - https://tools.ietf.org/html/rfc8628#section-3.4
+  // Device Access Token Response - https://tools.ietf.org/html/rfc8628#section-3.5
   let tokens;
-  let logged;
-
-  while (!done && !tokens) {
-    // Device Access Token Request https://tools.ietf.org/html/draft-ietf-oauth-device-flow-15#section-3.4
-    tokens = await client.grant({
-      grant_type: 'urn:ietf:params:oauth:grant-type:device_code',
-      device_code,
-    }).catch((err) => {
-      switch (err.error) {
-        case 'slow_down': // soft error, should re-try in the specified interval
-        case 'authorization_pending': // soft error, should re-try in the specified interval
-          if (!logged) {
-            console.log('\n\nauthorization pending ...');
-            logged = true
-          }
-          return wait(interval);
-          break;
-        case 'access_denied': // end-user declined the device confirmation prompt, consent or rules failed
-          console.error('\n\ncancelled interaction');
-          done = true;
-          break;
-        case 'expired_token': // end-user did not complete the interaction in time
-          console.error('\n\ndevice flow expired');
-          done = true;
-          break;
-        default:
-          if (err.name === 'OpenIdConnectError') {
-            console.error(`\n\nerror = ${err.error}; error_description = ${err.error_description}`);
-            done = true;
-          } else {
-            throw err;
-          }
-      }
-    });
+  try {
+    tokens = await handle.poll()
+  } catch (err) {
+    switch (err.error) {
+      case 'access_denied': // end-user declined the device confirmation prompt, consent or rules failed
+        console.error('\n\ncancelled interaction');
+        break;
+      case 'expired_token': // end-user did not complete the interaction in time
+        console.error('\n\ndevice flow expired');
+        break;
+      default:
+        if (err instanceof OPError) {
+          console.error(`\n\nerror = ${err.error}; error_description = ${err.error_description}`);
+        } else {
+          throw err;
+        }
+    }
   }
 
   if (tokens) {
+    console.log('\n\nresult tokens', { ...tokens });
+
     // requests without openid scope will not contain an id_token
     if (tokens.id_token) {
-      await client.validateIdToken(tokens); // validate ID Token (mandatory claims and signature)
```

Comment:
Is `client` already doing the validation now?

Reply:
yes.

Reply:
https://github.com/panva/node-openid-client/blob/v4.2.3/lib/device_flow_handle.js#L72-L75